### PR TITLE
oop-programs_with_interesting_classes-easy-q5

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/programs_with_interesting_classes/easy/q5/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# oop-programs_with_interesting_classes-easy-q5
+
+- Parsons Question
+
+## Question Text
+
+Social media and the Internet has enabled people to communicate with each other through great distances. Make a Java
+program that mimics a social media platform called `Discord`.
```
Comment (Hedgemon4) [inline src/oop/programs_with_interesting_classes/easy/q5/QuestionText.md:8]:
```suggestion
program that mimics a social media platform called `Discord` and another class to test it using the requirements below:
```

File: src/oop/programs_with_interesting_classes/easy/q5/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    public void discordClassHasCorrectDeleteMessageMethod(String value1, String value2) throws Throwable {
+        Object[][] constructorArgs = {
+                {value1, String.class},
+                {value2, String.class}
+        };
+        Object classInstance = testClass.createInstance(constructorArgs);
+        String incorrectNumDays = "The %s method in the %s class does not correctly delete the %s attribute.";
+        testClass.callMethod(methodName, classInstance);
+        _assertEquals("", testClass.getFieldValue(classInstance, attributeName1),
+                String.format(incorrectNumDays, methodName, className, attributeName1));
```
Comment (Hedgemon4) [inline src/oop/programs_with_interesting_classes/easy/q5/MainTest.java:113]: I think we can combine these tests as this is a parsons question.

Comment (Hedgemon4) [reply src/oop/programs_with_interesting_classes/easy/q5/MainTest.java:113]: Same thing for the get/set methods below.

File: src/oop/programs_with_interesting_classes/easy/q5/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    }
+
+    @Test
+    @Tag("dependency1")
+    public void correctMainMethod() throws Throwable {
+        Clause[] clauses = {
+                new StringLiteral("Hello there From: Jeremy"),
+                new NewLine(),
+                new StringLiteral("The sky is blue From: Jane")
+        };
```
Comment (Hedgemon4) [inline src/oop/programs_with_interesting_classes/easy/q5/MainTest.java:276]: Please add a custom error message for this call to the main method.

File: src/oop/programs_with_interesting_classes/easy/q5/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        assertTrue(testClass.hasModifier(getAttributeName1, null, "public"), String.format(incorrectModifierMessage, getAttributeName1, className));
+        assertTrue(testClass.hasReturnType(getAttributeName1, null, String.class), String.format(incorrectReturnType, getAttributeName1, className));
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] getMethodModifiers = {"public"};
+        String incorrectGetMethodMessage = String.join(" ",
+                "Your", getAttributeName1, "does not correctly get the value of the", attributeName1, "attribute.");
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q5/MainTest.java:130]:
```suggestion
                "Your", getAttributeName1, "method does not correctly get the value of the", attributeName1, "attribute.");
```

File: src/oop/programs_with_interesting_classes/easy/q5/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        assertTrue(testClass.hasModifier(getAttributeName2, null, "public"), String.format(incorrectModifierMessage, getAttributeName2, className));
+        assertTrue(testClass.hasReturnType(getAttributeName2, null, String.class), String.format(incorrectReturnType, getAttributeName2, className));
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] getMethodModifiers = {"public"};
+        String incorrectGetMethodMessage = String.join(" ",
+                "Your", getAttributeName2, "does not correctly get the value of the", attributeName2, "attribute.");
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q5/MainTest.java:151]:
```suggestion
                "Your", getAttributeName2, "method does not correctly get the value of the", attributeName2, "attribute.");
```

File: src/oop/programs_with_interesting_classes/easy/q5/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                {value1, String.class},
+                {value2, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {updatedValue1, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setAttributeName1, "does not correctly set the value of the", attributeName1, "attribute.");
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q5/MainTest.java:175]:
```suggestion
                "Your", setAttributeName1, "method does not correctly set the value of the", attributeName1, "attribute.");
```

File: src/oop/programs_with_interesting_classes/easy/q5/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void discordClassHasCorrectSetUserMethod(String value1, String value2, String updatedValue1, String updatedValue2) throws Throwable {
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
+        String incorrectModifierMessage = "The %s method in the %s class has the wrong visibility modifier.";
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod(setAttributeName2, new Class<?>[]{String.class}), String.format(incorrectMethodDefinition, setAttributeName2, className));
+        assertTrue(testClass.hasModifier(setAttributeName2, new Class<?>[]{String.class}, "public"), String.format(incorrectModifierMessage, setAttributeName2, className));
+        assertTrue(testClass.hasReturnType(setAttributeName2, new Class<?>[]{String.class}, Void.TYPE), String.format(incorrectReturnType, setAttributeName2, className));
+        Object[][] arguments = {
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q5/MainTest.java:189]: what is this used for?

File: src/oop/programs_with_interesting_classes/easy/q5/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+                {value1, String.class},
+                {value2, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {updatedValue2, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setAttributeName2, "does not correctly set the value of the", attributeName2, "attribute.");
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q5/MainTest.java:199]:
```suggestion
                "Your", setAttributeName2, "method does not correctly set the value of the", attributeName2, "attribute.");
```

File: src/oop/programs_with_interesting_classes/easy/q5/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    @MethodSource("inputProvider")
+    @Tag("dependent1")
+    public void discordClassHasCorrectToStringMethod(String value1, String value2) throws Throwable {
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
+        String incorrectModifierMessage = "The %s method in the %s class has the wrong visibility modifier.";
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod("toString", null), String.format(incorrectMethodDefinition, "toString", className));
+        assertTrue(testClass.hasModifier("toString", null, "public"), String.format(incorrectModifierMessage, "toString", className));
+        assertTrue(testClass.hasReturnType("toString", null, String.class), String.format(incorrectReturnType, "toString", className));
+        Object[][] arguments = {
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q5/MainTest.java:214]: what is this used for?

File: src/oop/programs_with_interesting_classes/easy/q5/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+                {value1, String.class},
+                {value2, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] methodModifiers = {"public"};
+        String expected = value1 + " From: " + value2;
+        String incorrectToStringMessage = String.join(" ",
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q5/MainTest.java:221]: make this a variable at the top of the tests

File: src/oop/programs_with_interesting_classes/easy/q5/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+                "Your", className, "toString method does not return the correct String.");
+        Object output = testClass.callMethod("toString", methodModifiers, classInstance);
+        _assertEquals(expected, output, incorrectToStringMessage);
+    }
+
+    @Test
+    @Tag("dependency1")
+    public void correctMainMethod() throws Throwable {
+        Clause[] clauses = {
+                new StringLiteral("Hello there From: Jeremy"),
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q5/MainTest.java:231]: make these also variables

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
See Sam's review

**Pull request merged**